DOC: Include warning in np.resize() docs

Anner-deJong opened the pull request:
I find it counterintuitive that resizing a numpy array does NOT consider dimensions independent, but simply takes the first N elements of the input array that fit into the output array. I was expecting something more like `cv2.resize()` where axes are treated separately, and some form of interpolation is applied to each dimension.

numpy resizes an old 5x5x2 array into a new array with new = np.resize(old, (4, 4, 2)) by simply copying the first 4 * 4 * 2 entries from old into new. (e.g. old[0, 4, :] becomes new[1, 0, :])
This is a problem if we would like simple interpolation, or if different axes represent different entities. (If the above example is an image, the top right most pixel would get warped all the way to the left, which is most likely not the intention)

In response to my issue #11478: 'resize() is unintuitive by not treating axes separately?'
A warning in the docs might prevent some confusion.

For a discussion on the issue and a possible solution, see <https://github.com/numpy/numpy/issues/11478>

Anner-deJong committed:
include warning in np.resize() docs

I find it counterintuitive that resizing a numpy array does NOT consider dimensions independent, but simply takes the first N elements of the input array that fit into the output array. I was expecting something more like `cv2.resize()` where axes are treated separately, and some form of interpolation is applied to each dimension.

numpy resizes an old 5x5x2 array into a new array with new = np.resize(old, (4, 4, 2)) by simply copying the first 4 * 4 * 2 entries from old into new. (e.g. old[0, 4, :] becomes new[1, 0, :])
This is a problem if we would like simple interpolation, or if different axes represent different entities. (If the above example is an image, the top right most pixel would get warped all the way to the left, which is most likely not the intention)

In response to my issue #11478: 'resize() is unintuitive by not treating axes separately?'
A warning in the docs might prevent some confusion.

For a discussion on the issue and a possible solution, see <https://github.com/numpy/numpy/issues/11478>

mattip commented on numpy/core/fromnumeric.py:
```diff
@@ -1197,7 +1197,17 @@ def resize(a, new_shape):
+    Notes
+    -----
+    Warning: This functionality does **not** consider axes separately,
+    i.e. it does not apply interpolation/extrapolation of some sort.
```
remove "of some sort"

Anner-deJong committed:
Update resize notes according to mattip's comments

removed 'simple'

altered sentence describing which items are taken from `old` into `new`

therefor -> therefore

Anner-deJong commented on numpy/core/fromnumeric.py:
```diff
@@ -1197,7 +1197,17 @@ def resize(a, new_shape):
+    -----
+    Warning: This functionality does **not** consider axes separately,
+    i.e. it does not apply interpolation/extrapolation.
+    It fills the return array with the required number of elements,
```
Perhaps 'It _instead_ fills the...?

mattip replied:
In my mind "instead" implies there is some single (better?) alternative, when in fact there are many other alternatives for what `resize` could be doing.

If we were starting over we may have called this `realloc`, since that links back to the memory model this function implements. Over the 20 years since it was created `resize` has taken on another meaning. Perhaps it would be nice to work the word `realloc` into the docstring.

Anner-deJong replied:
Thanks for your comments, let me try to work realloc in the docs.
I meant 'instead' more as in 'this function is not performing interpolation, but INSTEAD doing this and this.' Not trying to make either functionality superior, just highlighting the difference.

Anner-deJong replied:
hmm I might not full understand what you mean with `realloc`, sorry. The returned array of `np.resize()` is a newly created object, not an in-place view right? (i.e changing the resized array does not change the original array.)
Thus `np.resize()` allocates new memory, but doesn't re-allocate anything?

mattip replied:
correct, but the memory layout is like a `realloc` or a `memcpy`, or `np.ravel(a)` which is actually used internally to implement this. I think what you have is an improvement, maybe we should just use it as-is.

Anner-deJong replied:
The fact `np.ravel()` is being called internally makes a lot of sense yeah, but I'm not sure if including that in the note will make the docs more readable.

If you think it's okay as-is, I hope this pr can get accepted. If not / in the meantime, I'll think about the `realloc`/`np.ravel()` inclusion issue more.

mattip commented on numpy/core/fromnumeric.py:
```diff
@@ -1197,7 +1197,17 @@ def resize(a, new_shape):
+    This functionality is therefore not suitable to resize images,
+    or data where each axis represents a separate and distinct entity.
+    
     Examples
```
a final nit: this empty line and one other have spaces, would be nice if you could remove them

Anner-deJong committed:
removed spaces. improved right side outlining

charris commented:
Thanks @Anner-deJong .

Anner-deJong commented:
@charris, thanks for merging.

Can I ask, how the general pipeline works for deployment? i.e. when will the online docs (currently v1.14.5) show this and other updates (commits merged to master branch)?
Does it get updated ever so often, or only once in a while with new big releases?

eric-wieser commented:
@Anner: The 1.14.5 docs will never show this, since this commit went into 1.16-dev.

You can see the latest `master` docs at http://www.numpy.org/devdocs/, which should be deployed via CI

Anner-deJong commented:
Hi @eric-wieser,

Thanks for your quick reply, and I am sorry for my lack of release/workflow understanding. Hope the following questions are not too much to ask of your time:
(If there is some proper documentation about release procedures, would you have a link? Can't seem to find it)

In my understanding, currently the 'main' version is 1.14.5, and the 'main' documentation, https://docs.scipy.org/doc/numpy/, refers to this version. Is this correct?
I guess I was wondering, out of curiosity and interest, how and when the 'main' available release becomes a new version? Like when will the version-agnostic-url 'https://docs.scipy.org/doc/numpy/' refer to another version?
(And like what happened to 1.15, if a regular download gives one 1.14 but we are all working on the 1.16 dev?)
